### DISCUSSION TITLE
Mouse buttons and Modifier keys

### DISCUSSION THREAD

**Discussion by Steve-Mcl** [node-red/designs#73] [Design Proposals]
# Intro
An examination of current Mouse buttons and Modifier keys

<br>
<br>

# Mouse Down on: `canvas` (Button: `0` = LMB, Button: `2` = RMB)
## Current
Button | Modifiers | Mouse operation | Result *(Current)*| 🚩 Issue 
-|-|-|-|-
0|`none`<br>`alt`    | `up` | clears selection |
0|`shift`   | `up` | nothing |
0|`shift-alt`   | `up` | nothing |
0|`none`    | `move-up` | clears selection starts a NEW lasso then selects lassoed nodes |
0|`shift`   | `move-up` | starts a lasso then adds lassoed nodes to selection | 
0|`alt`     | `move-up` | clears selection starts a NEW lasso then selects lassoed nodes | 🚩duplicate
0|`alt-shift` | `move-up` | starts a lasso then adds lassoed nodes to selection | 🚩duplicate
0|`ctrl`    | `up` | clears selection and shows QUICK ADD |
0|`ctrl-shift`  | `up` | clears selection and shows QUICK ADD | 🚩duplicate
0|`ctrl-alt`    | `up` | clears selection and shows QUICK ADD | 🚩duplicate
0|`ctrl-shift-alt`    | `up` | clears selection and shows QUICK ADD | 🚩duplicate
0|`ctrl`| `move-up` | clears selection, shows QUICK ADD & disappears | 🚩duplicate
0|`ctrl-shift`| `move-up` | clears selection, shows QUICK ADD & disappears | 🚩duplicate
0|`ctrl-alt`| `move-up` | clears selection, shows QUICK ADD & disappears | 🚩duplicate
0|`ctrl-shift-alt`    | `move-up` | clears selection, shows QUICK ADD & disappears | 🚩duplicate
2|`ctrl`    | `move-up` | begins slice mode then disconnects any wires crossed | 🚩button `2`
2|`ctrl-shift`    | `move-up` | begins slice mode then disconnects any wires crossed | 🚩duplicate, button `2`
2|`ctrl-alt`    | `move-up` | begins slice mode then disconnects any wires crossed | 🚩duplicate, button `2`
2|`ctrl-shift-alt`    | `move-up` | begins slice mode then disconnects any wires crossed | 🚩duplicate, button `2`
2|`shift`    | `move-up` | begins slice mode then add junctions any wires crossed | 🚩button `2`
2|`shift-alt`    | `move-up` | begins slice mode then add junctions any wires crossed | 🚩duplicate, button `2`

## Issues
* Duplicates with various modifier combinations
* Both `ctrl` and `shift` are additive in different circumstances
   * With `0`-`ctrl`-up, the QUICK ADD dialog is displayed (additive) 
   * With `0`-`shift`-`move-up`, adds lassoed nodes to selection (additive) 
* `ctrl` is both additive and subtractive
  * With `0`-`ctrl`-up, the QUICK ADD dialog is displayed (additive) 
  * With `2`-`ctrl`-`move-up`, wires are sliced (subtractive)  
* `alt` does not affect any operations (as if not considered) 
* `shift` sometimes does not affect the operations
* Button 2 has issues on FF Browser

## Proposal
Button | Modifiers | Mouse operation | Result *(Proposed)* | ✍️ Changed 
-|-|-|-|-
0|`none`<br>`alt`    | `up` | clears selection |
0|`shift`   | `up` | nothing | 
0|`shift-alt`   | `up` | nothing |
0|`none`    | `move-up` | clears selection starts a NEW lasso then selects lassoed nodes |
0|`shift`   | `move-up` | starts a lasso then adds lassoed nodes to selection |
0|`alt`     | `move-up` |  begins slice mode then disconnects any wires crossed | ✍️ 
0|`alt-shift` | `move-up` | begins slice mode then add junctions any wires crossed | ✍️
0|`ctrl`    | `up` | clears selection and shows QUICK ADD |
0|`ctrl-shift`  | `up` | *Nothing (future operation)* | ✍️ 
0|`ctrl-alt`    | `up` | *Nothing (future operation)* | ✍️
0|`ctrl-shift-alt`    | `up` |  *Nothing (future operation)* | ✍️ 
0|`ctrl`| `move-up` |  *Nothing (future operation)* | ✍️ 
0|`ctrl-shift`| `move-up` | *Nothing (future operation)* | ✍️ |
0|`ctrl-alt`| `move-up` | *Nothing (future operation)* | ✍️ |
0|`ctrl-shift-alt`    | `move-up` | *Nothing (future operation)* | ✍️ |
2|`ctrl`    | `move-up` | *Nothing* | ✍️ Avoid Button `2` (firefox/context issues) |
2|`ctrl-shift`    | `move-up` | *Nothing* | ✍️ Avoid Button `2` (firefox/context issues) |
2|`ctrl-alt`    | `move-up` | *Nothing* | ✍️ Avoid Button `2` (firefox/context issues) |
2|`ctrl-shift-alt`    | `move-up` | *Nothing* | ✍️ Avoid Button `2` (firefox/context issues) |
2|`shift`    | `move-up` | begins slice mode then add junctions any wires crossed |  Avoid Button `2` (firefox/context issues) |
2|`shift-alt`    | `move-up` | begins slice mode then add junctions any wires crossed |  Avoid Button `2` (firefox/context issues) |

## Reasoning
* Distinct operations per modifier combination
* Avoids overloading Button 2

<br>
<br>

# Mouse Down on: `node` (Button: `0` = LMB, Button: `2` = RMB)
## Current
Button | Modifiers | Mouse operation | Result *(Current)*| 🚩 Issue 
-|-|-|-|-
*TODO*

## Proposal
Button | Modifiers | Mouse operation | Result *(Proposed)* | ✍️ Changed 
-|-|-|-|-
*TODO*

<br>
<br>

# Mouse Down on: `junction` (Button: `0` = LMB, Button: `2` = RMB)
## Current
Button | Modifiers | Mouse operation | Result *(Current)*| 🚩 Issue 
-|-|-|-|-
*TODO*

## Proposal
Button | Modifiers | Mouse operation | Result *(Proposed)* | ✍️ Changed 
-|-|-|-|-
*TODO*

## Comparisons (other popular applications)
*TODO*

**Comment by knolleary** [comment]
To summarise it in a different way, here is the list of *documented* mouse actions. 

#### Mouse Actions on Workspace

In all cases `ctrl` and `cmd` are interchangeable - we don't don't distinguish between them so we do not need Mac specific logic.

Mouse Action | Action
--|--
`Click`                      | Clears Selection
`Click-Drag`            | Lasso Select
`Shift-Click-Drag`  | Lasso Select - add to selection
`Ctrl-Click`             | Open quick add dialog (clears selection)
`RHClick-Drag`      | Slice mode to remove wires
`Shift-RHClick-Drag` | Slice mode to insert junctions

#### Proposed Changes

Mouse Action | Action
--|--
~`Ctrl-Click-Drag`~              |  This current opens the quick-add dialog, but closes it again when the mouse is released. This should be fixed to behave the same as `Ctrl-Click`
~`RHClick-Drag`~ -> `Alt-Click-Drag`                | Slice mode to remove wires
~`Shift-RHClick-Drag`~ -> `Alt-Shift-Drag`                | Slice mode to insert junctions


The end result will be no actions using right-click.
